meta-mender-tegra/recipes-bsp/u-boot/u-boot-mender-tegra-vars.inc

```diff
@@ -0,0 +1,31 @@
+MENDER_UBOOT_AUTO_CONFIGURE = "0"
+# Use the mmcblk0boot1 partition for uboot environment (partition 2
+# in uboot) or user partition (0) on Nano with SDcard environment
+MENDER_UBOOT_CONFIG_SYS_MMC_ENV_PART = "${@'0' if (d.getVar('TEGRA_MENDER_UBOOT_ENV_IN_SPIFLASH') or '') != '1' and (d.getVar('TEGRA_SPIFLASH_BOOT') or '') == '1' else '2'}"
+
+# Calculate this offset by adding up the offsets of each partition preceeding the uboot_env partition in sdmmc_boot and aligning to the next
+# 4096 byte boundary, then subtracting 4 MiB (4194304) since the sdmmc_boot represents the combined boot0 and boot1 partitions
+# Please note the suggestions in the nvidia thread at https://devtalk.nvidia.com/default/topic/1063652/jetson-tx2/mmcblk0boot1-usage-at-address-4177408-and-u-boot-parameter-storage-space-availability/
+# regarding suggestions about locating this environment area in mmcblk0boot1 at 0x3BB000 (3911680)
+# We use the same offset for tegra210 eMMC. For tegra210 SPI/SD, we
+# locate the environment in the padding block, before the VER
+# partition when using SPI flash.  Otherwise, we use the location of
+# the ENV partition on the SDcard - 0xC00000 (12582912).
+MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_tegra186 = "3911680"
+MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_tegra210 = "${@'3866624' if (d.getVar('TEGRA_MENDER_UBOOT_ENV_IN_SPIFLASH') or '') == '1' else '12582912' if (d.getVar('TEGRA_SPIFLASH_BOOT') or '') == '1' else '3911680'}"
+MENDER_UBOOT_MMC_ENV_LINUX_DEVICE_PATH_tegra210 = "${@'/dev/mtdblock0' if (d.getVar('TEGRA_MENDER_UBOOT_ENV_IN_SPIFLASH') or '') == '1' else '/dev/mmcblk0' if (d.getVar('TEGRA_SPIFLASH_BOOT') or '') == '1' else mender_get_uboot_env_mmc_linux_device_path(d)}"
+
+# Optionally use SDcard instead of SPI flash on Jetson Nano dev kits.
+# Starting with meta-tegra zeus branch, Nano dev kits use the SPI flash
+# for the U-Boot environment (as NVIDIA does with L4T R32.3.1 and
+# later).
+#
+# However, for Mender it may be more convenient to use an SDcard
+# partition, so the Mender environment is more self-contained.
+# Just set:
+#
+# TEGRA_MENDER_UBOOT_ENV_IN_SPIFLASH = ""
```

Comment:
These instructions should also point to an updated flash layout XML file with the ENV partition allocated at the right location.

Reply:
I think that you mean that we should select a different flash layout xml based on this variable (and not just update the comment)? Do you happen to have a copy of this flash layout somewhere? I poked around some of the old branches but I couldn't find it offhand.

Reply:
Oh, right, I think I added this before the changeover to using the XML file for SDcard formatting.  We'll need to create one, I guess.

Reply:
OK so we need to make a new flash_mender-uboot-env-in-sdcard.xml which reserves uboot space at 12582912  (0xC00000) on the sdcard to match what we are setting up in https://github.com/OE4T/meta-mender-community/blob/15635b9d9ed8d9064d32a5e3763b355520e54359/meta-mender-tegra/recipes-bsp/u-boot/u-boot-mender-tegra-vars.inc#L15-L15 

Since the current GP1 partition is size 2097152
https://github.com/OE4T/meta-mender-community/blob/15635b9d9ed8d9064d32a5e3763b355520e54359/meta-mender-tegra/recipes-bsp/tegra-binaries/mender-custom-flash-layout/jetson-nano-qspi-sd/flash_mender.xml#L89-L92
We'd make a reserved section for uboot-storage just after the GP1 partition with total space 12582912-2097152=10,485,760 bytes before the start of the uboot environment and then  262144 bytes for uboot vars, so total size at least 10747904 (0xA40000).
Then we'd add dual APP paritions instead of just one.

Assuming I've got this correct, is there a specific reason you'd suggest leaving 10MB before the uboot variables?  Are you planning room for NVIDIA to add more goodies there for us or increase the size of GP1?  It looks like we could alternatively move the environment down to 2097152 (0x200000) and make the size 262144  (0x40000)

I also apologize but I still don't understand what this comment means:
` Otherwise, we use the location of the ENV partition on the SDcard - 0xC00000 (12582912).`

Do you suspect the zeus branch had a different layout?  I can dig if you think it's important.

Reply:
> Do you suspect the zeus branch had a different layout? I can dig if you think it's important.

That comment was based on these changes: https://github.com/mendersoftware/meta-mender-community/pull/119/files

which were based on a different config file for handling SDcard formatting. I don't think 0xC00000 is fixed in stone for the changes we're making now, just make the offset align with whatever the actual position is.

Reply:
I thought maybe it would be obvious once I got there, but it's not. Putting an ENV partition before APP moves APP to mmcblk0p2 which will have a ripple effect I'd rather avoid. You alluded to a gpt feature to physically locate a partition somewhere on the disk without respect to its logical number. Is this somehow supported in the tegraflash sdcard formatting? I don't have a nano so I can't try that, but the make-sdcard method ends up like this:

```
Command (? for help): i
Partition number (1-17): 1
Partition GUID code: 0FC63DAF-8483-4772-8E79-3D69D8477DE4 (Linux filesystem)
Partition unique GUID: 20E6445D-E0D3-44D8-9331-A9E5CEF48D92
First sector: 2048 (at 1024.0 KiB)
Last sector: 2559 (at 1.2 MiB)
Partition size: 512 sectors (256.0 KiB)
Attribute flags: 0000000000000000
Partition name: 'ENV'

Command (? for help): i
Partition number (1-17): 2
Partition GUID code: 0FC63DAF-8483-4772-8E79-3D69D8477DE4 (Linux filesystem)
Partition unique GUID: BC7AB0CE-2BAA-4668-A90C-55A7CC5AA6EF
First sector: 4096 (at 2.0 MiB)
Last sector: 13476607 (at 6.4 GiB)
Partition size: 13472512 sectors (6.4 GiB)
Attribute flags: 0000000000000000
Partition name: 'APP'
```

Then I thought maybe an alternative would be just allocating more space for the GP1 partition in the XML file, but like we see in the first method, the size specified for this partition gets ignored, again, at least by make-sdcard:

```
Partition GUID code: 0FC63DAF-8483-4772-8E79-3D69D8477DE4 (Linux filesystem)
Partition unique GUID: 89CD8BF6-20B7-4F2E-9FBB-CA7F202E637D
First sector: 2048 (at 1024.0 KiB)
Last sector: 13474559 (at 6.4 GiB)
Partition size: 13472512 sectors (6.4 GiB)
Attribute flags: 0000000000000000
Partition name: 'APP'
```

Reply:
GP1 isn't a real partition, it's just a placeholder for the GPT, and that has a fixed size.

If there is no partition defined with allocation_attribute `0x808` (fill-to-end), make-sdcard will automatically place the APP partition at the end of the card, but that's all it does, as far as index/layout management goes (matching the logic in the `jetson-disk-image-creator.sh` script in L4T.  

For Mender, we probably want the data partition to be at the end, since APP and APP_b have to have fixed sizes, so if you want to physically locate APP after ENV but still have it have index 1, make-sdcard will need to be updated to handle that.

It might just be easier to modify the U-Boot patches to look up ENV by name, rather than using a fixed offset from the start of the card.  The `fw_env.config` file could then use `/dev/disk/by-partlabel/ENV` as the device name.

Reply:
> The fw_env.config file could then use /dev/disk/by-partlabel/ENV as the device name.

I like this idea, then for sdcard formatting the `DEVICE_FILE` type is still used in the [device name parsing logic in libubootenv](https://github.com/sbabic/libubootenv/blob/23b30860c167df0c2bf1c0083592938475969d36/src/uboot_env.c#L150).  I think we could actually use this strategy for all u-boot cases except `TEGRA_MENDER_UBOOT_ENV_IN_SPIFLASH = "1"` where we need to use mtd?

Reply:
> I think we could actually use this strategy for all u-boot cases except TEGRA_MENDER_UBOOT_ENV_IN_SPIFLASH = "1" where we need to use mtd?

It only really applies to jetson-nano-qspi-sd, since it's the only one that is SDcard based and still uses U-Boot.  The default should be to keep the env in the SPI flash, since that's the L4T out-of-the-box default as of 32.4.3, so this ENV partition hackery should be treated as a special case.  For the eMMC-based devices, it's still the eMMC's boot1 area.

Reply:
I've committed 000e8d55e6a43cba075f38603ec41b630151fb90 to do this. This is completely untested, since I don't have a Nano, so please keep the fire extinguisher handy and don't blame me if you have to use it 🧑‍🚒

Reply:
I looked at this more closely this morning, and I retract my suggestion as it just cannot be made to work without some major changes to U-Boot itself.  It was a bad idea, and I'm sorry we wasted so much effort on it. :(

First, @kekiefer, that patch isn't quite right... `CONFIG_SYS_MMC_ENV_PART` refers to a _hard_ partition in an MMC, not a soft partition, so it must be zero for an SDcard.  The partition name is used for calculating the offset.  There's code in `env/mmc.c` for doing this by storing the name of the partition under the `/config` section of the U-Boot compiled-in device tree, which is why I thought this might work - at first glance, we could have used that.

However, that code *doesn't* support redundancy, and it hard-codes the offset as being from the end of the named partition.  (There's code for handling device tree entries for offset and redundant offset, but they are *not* used if a partition name is provided.  Sigh.)

Furthermore, U-Boot uses a `uint32` type for holding the offset value in its common code, which means that it must be less than 4GiB from the front of the device (_not_ the front of the soft partition).  So that pretty much means that the ENV partition *would* have to be first on the SDcard for our use case, since it's likely that folks would use cards larger than 8GB - and even with an 8GB card, we'd be pushing our luck on assuming that the starting offset for ENV would be under the 4GiB upper limit.

I now recommend just forgetting about trying to support this going forward.  The amount of hackery involved in making it work is just too high to be maintainable.  I'm now going to go try and glue back on the hair I pulled out this morning trying to make this work.

Reply:
Thanks for looking at this again @madisongh.

I'm all for simplifying `jetson-nano-qspi-sd` support and supporting in only one way.  I read back through some of the old discussion about this to try to refresh my memory about this feature, starting with [this PR](https://github.com/mendersoftware/meta-mender-community/pull/119#issuecomment-571519925).  As I understand it the main drawback of only supporting `TEGRA_MENDER_UBOOT_ENV_IN_SPIFLASH = ''1"` would be that if someone uses `dosdcard.sh` and doesn't tegraflash, their uboot environment will still contain stale data from the last tegraflash, which might mean they start by booting APP partition B instead of A (which shouldn't really hurt anything, just might be confusing), or if they tegraflashed with an old image it might mean they have some incompatability in their uboot environment which would need to be tracked down.  If I understand correctly, I think we can avoid these issues with an "if in doubt, tegraflash" mantra for anyone starting out with this platform or for anyone switching yocto branches.  Also upgrades across yocto branches on this platform are probably a lower priority since it's not meant for production anyway, so it's more likely a tegraflash requirement would be acceptable in this scenario.

Reply:
I've dropped support for `TEGRA_MENDER_UBOOT_ENV_IN_SPIFLASH` in 2a08c75. I'll squash this commit and reorder things after I get a thumbs up that nano is good to go.